### ISSUE TITLE
nan in fragments after rasterizer

### ISSUE THREAD

**Issue by AsaphLightricks** [facebookresearch/pytorch3d#460]
[meshes_nan.zip](https://github.com/facebookresearch/pytorch3d/files/5584033/meshes_nan.zip)
## NaN values after rasterizer
During the training of my network I render meshes. In some batches of meshes, after i pass them through the rasterizer, a single pixel of a single output image will have NaN values in the bary_coords and the zbuf.

## Instructions To Reproduce the Issue:
Unzip 'meshes_nan.zip' and run `python show_mesh_nan.py`

contents of show_mesh_nan.py:

```
from pytorch3d.renderer import (
    FoVPerspectiveCameras,
    MeshRasterizer,
    look_at_view_transform,
    RasterizationSettings,
)
from pytorch3d.structures import Meshes
import torch

device = 'cpu'

PI = 3.1415926
Z_NEAR = -10.0
Z_FAR = -3000.0
SENSOR_SIZE = 24.0
FOCAL_LENGTH = 40.0

raster_settings = RasterizationSettings(
    image_size=512,
    blur_radius=1e-6,
    faces_per_pixel=1,
    clip_barycentric_coords=True,
    cull_backfaces=True,
    perspective_correct=True,
)
rasterizer = MeshRasterizer(raster_settings=raster_settings)

R, T = look_at_view_transform(at=[[0, 0, -1]], eye=[[0, 0, 0]], device=device)
half_sensor = torch.tensor(SENSOR_SIZE / 2)
fov = 2 * torch.atan(half_sensor / FOCAL_LENGTH) / PI * 180

one_camera = FoVPerspectiveCameras(
    fov=fov, degrees=True, znear=Z_NEAR, zfar=Z_FAR, R=R, T=T, device=device
)

meshes = torch.load("my_meshes.pt", map_location=device)
fragments = rasterizer(meshes[8], cameras=one_camera)

print(torch.where(torch.isnan(fragments.pix_to_face)))
print(torch.where(torch.isnan(fragments.bary_coords)))
print(torch.where(torch.isnan(fragments.zbuf)))
print(torch.where(torch.isnan(fragments.dists)))
```

and my output is:
```
(tensor([], dtype=torch.int64), tensor([], dtype=torch.int64), tensor([], dtype=torch.int64), tensor([], dtype=torch.int64))
(tensor([0, 0]), tensor([158, 158]), tensor([407, 407]), tensor([0, 0]), tensor([1, 2]))
(tensor([0]), tensor([158]), tensor([407]), tensor([0]))
(tensor([], dtype=torch.int64), tensor([], dtype=torch.int64), tensor([], dtype=torch.int64), tensor([], dtype=torch.int64))

Process finished with exit code 0
```

I use pytorch3d 0.3.0 with torch 1.6.0
problem also appears with pytorch 0.2.5 and torch 1.5.0

Can you help?

**Comment by gkioxari** [comment]
Your `znear` and `zfar` parameters seem totally off and likely what is causing the problem. In PyTorch3D we assume that +Z goes from you towards the monitor. Negative values for `znear` and `zfar` means that your mesh is behind the camera?

**Comment by AsaphLightricks** [comment]
@gkioxari thanks for responding.
I'm aware of the PyTorch3D convention. However if you'll see in my code, I set the 'at' vector of the camera to [0, 0, -1] meaning i turn the camera to look toward the -Z axis. I'm use to working with OpenGL so i 'override' the pytorch3d convention by turning the camera. Once i do that, the znear and zfar values make sense. Anyway, i get good images for most of my rendering, so i don't think that's the issue

**Comment by AsaphLightricks** [comment]
Ok, i looked the code and i realized something. The rasterizer doesn't really use the znear and zfar at the moment, from `rasterizer.py`:
```
# NOTE: Retaining view space z coordinate for now.
# TODO: Revisit whether or not to transform z coordinate to [-1, 1] or
# [0, 1] range.
verts_view = cameras.get_world_to_view_transform(**kwargs).transform_points(
            verts_world
    )
verts_screen = cameras.get_projection_transform(**kwargs).transform_points(
            verts_view
    )
verts_screen[..., 2] = verts_view[..., 2]
```
the z values that were mapped to range [0, 1] (using znear and zfar) are overwritten by the view space z values (which indeed are positive)

Then in my own code when i use the shader i pass it the absolute values of znear and zfar:
```
images = self.shader(
            fragments, meshes, sigmas=sigmas, znear=abs(self.z_near), zfar=abs(self.z_far)
        )
```
which conforms the PyTorch3D convention.

In summary, yes, i used negative znear and zfar values instead of positive, and I 'accidentally' fixed my error by passing the absolute values to the shader, probably because i wasn't getting correct images, which is why I do end up with nice images most of the time.

However, my problem is in the Rasterizer, which doesn't really use the znear and zfar values, so that shouldn't be the root cause.

**Comment by gkioxari** [comment]
The rasterizer doesn't use the `znear` and `zfar` but the camera that transforms the points before they are being passed through the rasterizer does use them, so they can affect the output of the rasterizer

**Comment by gkioxari** [comment]
I can't understand how you are defining your cameras and transforms. Here is a working example for your mesh that produces the final output 

![mesh](https://user-images.githubusercontent.com/4369065/100133903-39f1de80-2e3c-11eb-8d60-9fb746112fae.png)

```python

    device = torch.device("cuda:1")
    meshes = torch.load("/tmp/output/my_meshes.pt", map_location=device)
    mesh = meshes[8]

    verts = mesh.verts_list()[0].detach()
    faces = mesh.faces_list()[0].detach()

    # transform to pytorch3d world system
    verts[:, 0] *= -1.0
    verts[:, 2] *= -1.0
    tex = TexturesVertex([torch.ones_like(verts)])
    mesh = Meshes(verts=[verts], faces=[faces], textures=tex)

    raster_settings = RasterizationSettings(
        image_size=512,
        blur_radius=1e-6,
        faces_per_pixel=1,
        clip_barycentric_coords=True,
        cull_backfaces=True,
        perspective_correct=True,
    )
    camera = PerspectiveCameras(device=device, focal_length=2.5)

    rasterizer = MeshRasterizer(cameras=camera, raster_settings=raster_settings)
    shader = HardGouraudShader(device=device, cameras=camera)

    fragments = rasterizer(mesh)
    image = shader(fragments, mesh)

    assert torch.isfinite(fragments.pix_to_face).all()
    assert torch.isfinite(fragments.dists).all()
    assert torch.isfinite(fragments.bary_coords).all()
    assert torch.isfinite(fragments.zbuf).all()

    filename = "/tmp/output/mesh.png"
    rgb = image[0, ..., :3].detach().squeeze().cpu()
    Image.fromarray((rgb.detach().cpu().numpy() * 255).astype(np.uint8)).save(filename)
```
